### DISCUSSION TITLE
how are multiple IO stages treated in Amanzi?

### DISCUSSION THREAD

**Discussion by ecoon** [amanzi/amanzi#731] [General]
@lipnikov  In the tpetra merge process, I'm seeing some changes in how HDF5_MPI deals with "io sets" or "xmlStep" -- effectively this seems to be an internal counter at which point the xml file names change.

Can you explain this a bit?  I believe this is coming from an Amanzi MPC feature around "multiple simulation stages in the same run."  Do these stages share the same h5 files?  The same mesh h5 and xmf?  Is there a reason there is persistent information stored inside the HDF5_MPI object between these stages?

It would make more sense to me to just make a new Output object for each stage, but I'm not sure what you're trying to share across stages.

**Comment by jd-moulton** [comment]
@ecoon @lipnikov I can add a bit of background/motivation here.  The problem this feature is solving is with a single run that involves a spin-up step that may not have all processes present.  Specifically, with our EM problems there's often a hydrology spin-up phase, and then we turn on transport (or reactive transport), for the transient phase.    VisIT and ParaView don't like that all fields are not present in the h5 file for all times in the simulation (i.e., they crash or refuse to do the vis). So the spin-up and transient phases were put in separate h5 files.   In general, there could be many phases with different forcing and slightly different configurations, so the idea was that each phase would get a separate h5 file, and file names would include the phase number of the simulation to keep things straight.  

You could argue that these could all be separate runs, but the workflow the EM folks wanted/needed was for this to be setup as a single run.  I agree that a new Output object could be used for each stage and might simplify things.   The content of the output for each stage is the default for the set of processes present, modulo any black/white list controls.

**Comment by lipnikov** [comment]
@ecoon I can add that stages require to write same metadata to multiple files. Potentially, with  a finer i/o API a cycle driver can do it. Since this may affect ATS, require many changes, the internal bookkeeping strategy was implemented. It recognizes that State is different (xmlStep != xmlStep_prev) and launches a new io-set.
